### PR TITLE
docs: ray distributed debugger

### PR DESCRIPTION
**Pull request by saihaj**
## Why are these changes needed?

Introduce ray distributed debugger docs and also the form for downloading the debugger.

CleanShot 2024-06-19 at 00 11 20@2x

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+.. image:: ./images/attach-paused-task.gif
+    :align: center
+
+- **Use the VSCode Debugger**: Debug your Ray application just as you would when developing locally.
+
+
+Post-Mortem Debugging
+=====================
+
+Continuing from the previous section, let's delve into Post-Mortem debugging. This feature becomes essential when Ray tasks encounter unhandled exceptions. In such cases, Ray automatically freezes the failing task, awaiting attachment by the Ray Debugger. This functionality empowers you to thoroughly investigate and inspect the program's state at the time of the error.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:129]:
```suggestion
Use post-mortem debugging when Ray tasks encounter unhandled exceptions. In such cases, Ray automatically freezes the failing task, awaiting attachment by the Ray Debugger. This functionality allows you to thoroughly investigate and inspect the program's state at the time of the error.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+- **Run a Ray Task Raised Exception**: Run the same `job.py` created above with an additional argument raise-exception.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:131]:
```suggestion
Run a Ray Task Raised Exception
~~~~~~~~~~~~~~~~~~~~~~~~
 
Run the same `job.py` created above with an additional argument raise-exception.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    .. code-block:: bash
+
+        python job.py raise-exception
+
+- **Attach to Paused Tasks**:
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:137]:
```suggestion
Attach to a paused task
~~~~~~~~~~~~~~~~~~

```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  - The task will be frozen once the exception is thrown.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:138]:
```suggestion
When the app throws an exception:

  - The debugger freezes the task.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  - The terminal will clearly indicate when a task is paused and waiting for the debugger to attach.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:139]:
```suggestion
  - The terminal clearly indicates when the debugger pauses a task is paused and waits for the debugger to attach.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  - The paused tasks will be listed in the Ray Debugger extension.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:140]:
```suggestion
  - The paused task is listed in the Ray Debugger extension.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  - Click on a paused task to attach the VSCode debugger.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:141]:
```suggestion
  - You can click on a paused task to attach the VS Code debugger.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+.. image:: ./images/post-moretem.gif
+    :align: center
+
+- **Use the VSCode Debugger**: Debug your Ray application just as you would when developing locally.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:146]:
```suggestion
Use the VSCode Debugger
~~~~~~~~~~~~~~~~~~~~

Debug your Ray app just as you would when developing locally.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+Feedback
+=========
+
+Join `#ray-debugger <https://ray-distributed.slack.com/archives/C073MPGLAC9>`_ channel on the Ray slack channel to get help
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:151]:
```suggestion
Join the `#ray-debugger <https://ray-distributed.slack.com/archives/C073MPGLAC9>`_ channel on the Ray Slack channel to get help
```

**Review by angelinalg — COMMENTED** [review]
That's a lot of suggestions, I know. Could you address those and then let me take another pass? It's difficult to envision what it's going to look like until then. Thanks!

**Review by aslonnie — COMMENTED** [review]
please properly rebase?

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+    conda create -n myenv python=3.9
+    conda activate myenv
+    pip install "ray[default]" debugpy
+
+
+Start a Ray cluster
+~~~~~~~~~~~~~~~~~~~
+
+Run `ray start --head` to start a Ray Cluster.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:44]:
```suggestion
Run `ray start --head` to start a Ray cluster.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+
+Register the cluster
+~~~~~~~~~~~~~~~~~~~~
+
+Add the Ray cluster `IP:PORT` to the cluster list. The default `IP:PORT` is `127.0.0.1:8265`, and you can change it when starting a new cluster. Make sure your current machine can access the IP and port.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:50]:
```suggestion
Find and click the Ray extension in the VS Code left side nav. Add the Ray cluster `IP:PORT` to the cluster list. The default `IP:PORT` is `127.0.0.1:8265`. You can change it when you start the cluster. Make sure your current machine can access the IP and port.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+- Click on the paused task to attach the VS Code debugger.
+
+.. image:: ./images/attach-paused-task.gif
+    :align: center
+
+
+Use the VS Code debugger
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Debug your Ray app just as you would when developing locally.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:91]:
```suggestion
Debug your Ray app as you would when developing locally.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+Attach to a paused task
+~~~~~~~~~~~~~~~~~~~~~~~
+
+When debugger hits a breakpoint
+
+- The task enters a paused state.
+- The terminal clearly indicates when the debugger pauses a task and waits for the debugger to attach.
+- The paused task is listed in the Ray Debugger extension.
+- Click on the paused task to attach the VS Code debugger.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:82]:
```suggestion
- Click the play icon next to the name of the paused task to attach the VS Code debugger.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+- Click the play icon next to the name of the paused task to attach the debugger and start debugging. 
+
+.. image:: ./images/post-moretem.gif
+    :align: center
+
+
+Use the VS Code debugger
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Debug your Ray app just as you would when developing locally.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:126]:
```suggestion
Debug your Ray app as you would when developing locally.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+    python job.py raise-exception
+
+
+Attach to paused task
+~~~~~~~~~~~~~~~~~~~~~
+
+When the app throws an exception:
+
+- The debugger freezes the task.
+- The terminal clearly indicates when the debugger pauses a task is paused and waits for the debugger to attach.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:115]:
```suggestion
- The terminal clearly indicates when the debugger pauses a task and waits for the debugger to attach.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+    <div style="position: relative; height: 0; overflow: hidden; max-width: 100%; height: auto;">
+        <iframe width="560" height="315" src="https://www.youtube.com/embed/EiGHHUXL0oI" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>
+    </div>
+
+
+Get started
+===========
+
+Click `here <https://www.anyscale.com/blog/ray-distributed-debugger?utm_source=ray_docs&utm_medium=docs&utm_campaign=promotion#download-for-free>`_ to download the Ray Debugger extension for free.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:25]:
```suggestion
`Download <https://www.anyscale.com/blog/ray-distributed-debugger?utm_source=ray_docs&utm_medium=docs&utm_campaign=promotion#download-for-free>`_ the Ray Debugger extension.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+====
+
+.. raw:: html
+
+    <div style="position: relative; height: 0; overflow: hidden; max-width: 100%; height: auto;">
+        <iframe width="560" height="315" src="https://www.youtube.com/embed/EiGHHUXL0oI" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>
+    </div>
+
+
+Get started
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:22]:
```suggestion
Download the extension
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+===========
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:23]:
```suggestion
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+Demo
+====
+
+.. raw:: html
+
+    <div style="position: relative; height: 0; overflow: hidden; max-width: 100%; height: auto;">
+        <iframe width="560" height="315" src="https://www.youtube.com/embed/EiGHHUXL0oI" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>
+    </div>
+
+
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:21]:
```suggestion
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+Get started
+===========
+
+Click `here <https://www.anyscale.com/blog/ray-distributed-debugger?utm_source=ray_docs&utm_medium=docs&utm_campaign=promotion#download-for-free>`_ to download the Ray Debugger extension for free.
+
+
+Setup environment
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:28]:
```suggestion
Set up the environment
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+~~~~~~~~~~~~~~~~~
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:29]:
```suggestion
~~~~~~~~~~~~~~~~~~
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+~~~~~~~~~~~~~~~~
+
+Start running your Ray app.
+
+.. code-block:: bash
+
+    python job.py
+
+
+Attach to a paused task
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:74]:
```suggestion
Attach to the paused task
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~
+
+When debugger hits a breakpoint
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:77]:
```suggestion
When the debugger hits a breakpoint:
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+- The task enters a paused state.
+- The terminal clearly indicates when the debugger pauses a task and waits for the debugger to attach.
+- The paused task is listed in the Ray Debugger extension.
+- Click on the paused task to attach the VS Code debugger.
+
+.. image:: ./images/attach-paused-task.gif
+    :align: center
+
+
+Use the VS Code debugger
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:88]:
```suggestion
Start debugging
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Run the same `job.py` file with an additional argument to raise an exception.
+    
+.. code-block:: bash
+
+    python job.py raise-exception
+
+
+Attach to paused task
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:109]:
```suggestion
Attach to the paused task
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+~~~~~~~~~~~~~~~~~~~~~
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:110]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+- The debugger freezes the task.
+- The terminal clearly indicates when the debugger pauses a task is paused and waits for the debugger to attach.
+- The paused task is listed in the Ray Debugger extension.
+- Click the play icon next to the name of the paused task to attach the debugger and start debugging. 
+
+.. image:: ./images/post-moretem.gif
+    :align: center
+
+
+Use the VS Code debugger
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:123]:
```suggestion
Start debugging
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 28]
```diff
@@ -0,0 +1,132 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Debug your Ray app just as you would when developing locally.
+
+
+Feedback
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:129]:
```suggestion
Share feedback
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 29]
```diff
@@ -0,0 +1,132 @@
...
+========
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:130]:
```suggestion
==============
```

**Review by angelinalg — COMMENTED** [review]
Just waiting for the CI build for a final check before approving. Thanks for addressing my comments!

**Comment by angelinalg** [comment]
Could you add a `Next steps` section at the end that refers to this doc:https://docs.ray.io/en/latest/ray-observability/user-guides/debug-apps/ray-debugging.html?

Maybe you can say something like: `For more guidance on debugging distributed apps in Ray, see <link>.`

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 30]
```diff
@@ -0,0 +1,137 @@
...
+Share feedback
+==============
+
+Join the `#ray-debugger <https://ray-distributed.slack.com/archives/C073MPGLAC9>`_ channel on the Ray Slack channel to get help.
+
+
+Next steps
+==========
+
+For guidance on debugging distributed apps in Ray, see :doc:`General debugging <./user-guides/debug-apps/general-debugging>`.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:136]:
```suggestion
- For guidance on debugging distributed apps in Ray, see :doc:`General debugging <./user-guides/debug-apps/general-debugging>`.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 31]
```diff
@@ -0,0 +1,137 @@
...
+For tips on using the Ray debugger, see :doc:`Ray debugging <./user-guides/debug-apps/ray-debugging>`.
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:137]:
```suggestion
- For tips on using the Ray debugger, see :doc:`Ray debugging <./user-guides/debug-apps/ray-debugging>`.
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 32]
```diff
@@ -0,0 +1,137 @@
...
+
+The Ray Distributed Debugger is a VS Code extension that streamlines the debugging process with an interactive debugging experience. The Ray Debugger enables you to:
+
+- **Break into remote tasks**: Set a breakpoint in any remote task. A breakpoint pauses execution and allows you to connect with VS Code for debugging.
+- **Post-mortem debugging**: When Ray tasks fail with unhandled exceptions, Ray automatically freezes the failing task and waits for the Ray Debugger to attach, allowing you to inspect the state of the program at the time of the error.
+
+Ray Distributed Debugger abstracts the complexities of debugging distributed systems for you to debug Ray applications more efficiently, saving time and effort in the development workflow.
+
+
+Demo
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:12]:
```suggestion
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 33]
```diff
@@ -0,0 +1,137 @@
...
+====
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:13]:
```suggestion
```

File: doc/source/ray-observability/ray-distributed-debugger.rst (hunk 32 above)
Comment (angelinalg) [reply doc/source/ray-observability/ray-distributed-debugger.rst:12]: Let's remove this heading. It's not necessary.

**Review by angelinalg — APPROVED** [review]
Could you accept these last suggestions? I'll approve in anticipation.

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 34]
```diff
@@ -0,0 +1,135 @@
...
+
+
+
+.. raw:: html
+
+    <div style="position: relative; height: 0; overflow: hidden; max-width: 100%; height: auto;">
+        <iframe width="560" height="315" src="https://www.youtube.com/embed/EiGHHUXL0oI" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>
+    </div>
+
+Download the extension
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:19]:
```suggestion
```

File: doc/source/ray-observability/ray-distributed-debugger.rst [hunk 35]
```diff
@@ -0,0 +1,135 @@
...
+
```
Comment (angelinalg) [inline doc/source/ray-observability/ray-distributed-debugger.rst:20]:
```suggestion

```

Comment (angelinalg) [reply doc/source/ray-observability/ray-distributed-debugger.rst:20]: Missed this extraneous text last time.

File: doc/source/ray-observability/doc_code/ray-distributed-debugger.py [hunk 36]
```diff
@@ -0,0 +1,29 @@
...
+    runtime_env={
+        "env_vars": {"RAY_DEBUG": "1"},
+    }
+)
+
+
+@ray.remote
+def my_task(x):
+    y = x * x
+    breakpoint()  # Add a breakpoint in the ray task.
```
Comment (angelinalg) [inline doc/source/ray-observability/doc_code/ray-distributed-debugger.py:15]:
```suggestion
    breakpoint()  # Add a breakpoint in the Ray task.
```

Comment (angelinalg) [reply doc/source/ray-observability/doc_code/ray-distributed-debugger.py:15]: I think this correction got lost in the other suggestions.

**Review by angelinalg — APPROVED** [review]

**Pull request merged**